### DISCUSSION TITLE
⚡️I tried to reduce the size of the extension through the rollup's external config, but I failed!

### DISCUSSION THREAD

**Discussion by seymoe** [directus/directus#18015] [Archive]
When I write and use extensions, the extension's size makes the loading speed of the website very slow because the extension relies on many third-party npm packages, they are bundled into `index.js`.

I try to solve this problem, I found that directus have this config, only these packages can be used as external dependencies.

```
export const APP_SHARED_DEPS = ['@directus/extensions-sdk', 'vue', 'vue-router', 'vue-i18n', 'pinia']
```
I find a way to add external option from [#7332](https://github.com/directus/directus/pull/7332) , I add a file named `extension.config.mjs` in my extension, and then `npm run build`, it's worked, the bundle size reduced volume from 1.5M to 900kb, Good!!

```
function externalPlugin(options = {}) {
  return {
    name: "external-plugin",
    options(inputOptions) {
      const externals = inputOptions.external;
      const customExternals = options.externals;
      inputOptions.external = Array.isArray(customExternals)
        ? [...externals, ...customExternals]
        : externals;
    }
  };
}

export default {
  plugins: [
    externalPlugin({
      externals: ['lodash', 'date-fns', '@revolist/vue3-datagrid', '@revolist/revogrid-column-select', '@revolist/revogrid-column-numeral', '@province-city-china/city', 'echarts', 'echarts-stat', 'vue-echarts', 'numeral']
    })
  ]
}
```

Unfortunately, when i run the directus server and load extension ...

```
"lodash" is imported by "extensions/modules/directus-extension-xxx-module/index.js", but could not be resolved – treating it as an external dependency.
"vue-echarts" is imported by "extensions/modules/directus-extension-xxx-module/index.js", but could not be resolved – treating it as an external dependency.
"date-fns" is imported by "extensions/modules/directus-extension-xxx-module/index.js", but could not be resolved – treating it as an external dependency.
"@province-city-china/city" is imported by "extensions/modules/directus-extension-xxx-module/index.js", but could not be resolved – treating it as an external dependency.
...
```

And than I found directus do `generateExtensionBundle` again before server start, It also use `APP_SHARED_DEPS` config, and also used in the `@directus/app` [https://sourcegraph.com/github.com/directus/directus@953c2f9/-/blob/app/vite.config.js?L83](https://sourcegraph.com/github.com/directus/directus@953c2f9/-/blob/app/vite.config.js?L83)

So, Is there any other way to achieve this? 
I don't know much about directus, exploring this step is already my limit!
help me Pls!!

**Comment by rijkvanzanten** [comment]
The app shared deps listed in that const are dependencies that the app makes globally available; therefore, rollup is able to treat them as externals and ignore them in the bundle. However, other dependencies you introduce to the extension, like `echarts` (1MB), aren't globally available, so have to be bundled into the extension. If you want to use echarts (which is a huge library), you have to account for the extension being big. Not much you can do about that!

**Comment by gbicou** [comment]
I have pretty much the same issue with external dependencies in an extension, to work around this limitation I created the pr #18084 wich resolve node dependencies when bundling extensions and serve all rollup chunks.
